docs/source/framework/pytorch_integration/writing_layers.rst

```diff
@@ -0,0 +1,280 @@
+    import torch
+    lang = """
+    def matmul(float(M,N) A, float(N,K) B) -> (output) {
+      output(i, j) +=! A(i, kk) * B(kk, j)
+    }
+    def abs(float(M, N) A) -> (O1) {
+      O1(m, n) = fabs(A(m, n))
+    }
+    """
+    matmul = tc.define(lang, name="matmul")
+    mat1, mat2 = torch.randn(3, 4).cuda(), torch.randn(4, 5).cuda()
+    out = matmul(mat1, mat2)
+
+    abs = tc.define(lang, name="abs")
+    A = torch.randn(3, 4).cuda()
+    out = abs(A)
+
+
+Writing layers with scalars
+---------------------------
+
+If you have an operation that requires a constant scalar value for bounds inference,
+for example, kernel or stride in case of convolution operation, we need to pass
+the TC with the substituted scalar value because right now, we don't support using
+scalars for bound inference. The substitution can be done in two ways and users can
+adopt whatever feels more convenient.
+
+* **Option 1**: Pass a constants dictionary to the :code:`tc.define` call. An example for how to do this easily is below:
+
+.. code-block:: python
+
+    import tensor_comprehensions as tc
+    import torch
+    lang = """
+    def avgpool(float(B, C, H, W) input) -> (output) {{
+        output(b, c, h, w) += input(b, c, h * {sH} + kh, w * {sW} + kw) where kh in 0:{kH}, kw in 0:{kW}
+    }}
+    """
+    avgpool = tc.define(lang, name="avgpool", constants={"sH":1, "sW":1, "kH":2, "kW":2})
+    inp = torch.ones(32, 3, 10, 10).cuda()
+    out = avgpool(inp)
+
+.. note::
+
+    In python, the formatting of strings requires usage of :code:`{{...}}`. Hence
+    the above example uses these brackets. You only need to do this if your TC
+    consists of scalars.
+
+* **Option 2**: Format the string using python regex. An example below:
+
+.. code-block:: python
+
+    import tensor_comprehensions as tc
+    import torch
+    import re
+    LANG="""
+    def avgpool(float(B, C, H, W) input) -> (output) {
+        output(b, c, h, w) += input(b, c, h * <sh> + kh, w * <sw> + kw) where kh in 0:<kH>, kw in 0:<kW>
+    }
+    """
+    sH, sW, kH, kW = 1, 1, 2, 2
+    LANG = re.sub('<sh>', str(sH), LANG)
+    LANG = re.sub('<sw>', str(sW), LANG)
+    LANG = re.sub('<kH>', str(kH), LANG)
+    LANG = re.sub('<kW>', str(kW), LANG)
+    avgpool = tc.define(LANG, name="avgpool")
+    inp = torch.ones(1, 1, 4, 4).cuda()
+    out = avgpool(inp)
+
+
+Manually Injecting external CUDA code
+-------------------------------------
+
+If you have an external efficient CUDA code that you want to use rather than
+the CUDA code that TC generates, you can inject your code easily. For this,
+you need to create a string which has the CUDA code you want to inject and you
+need to pass the name of the kernel and the CUDA code string to the :code:`tc.define`
+call. For example:
+
+.. code-block:: python
+
+    import tensor_comprehensions as tc
+    import torch
+    lang = """
+    def add(float(N) A, float(N) B) -> (output) {
+        output(i) = A(i) + B(i) + 1
```

Comment:
same side effect of copy-pasta as in #72 : your add function adds an extra 1 that your my_add function does not

Reply:
you are reviewing a stacked PR, this is part of base PR #72